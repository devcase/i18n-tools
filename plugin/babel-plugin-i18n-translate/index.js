"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _ignoreAstPath = _interopRequireDefault(require("../ignore-ast-path"));

var _defineKey = _interopRequireDefault(require("../define-key"));

var _fluent = require("fluent");

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _helperPluginUtils = require("@babel/helper-plugin-utils");

var _core = require("@babel/core");

var _default = (0, _helperPluginUtils.declare)(function (api, options) {
  var locale = options.locale || "";

  var translationsFile = _path.default.join('i18n', 'translations' + (locale ? "." + locale : "") + '.ftl');

  var translationFileContents = _fs.default.readFileSync(translationsFile, {
    encoding: "UTF-8"
  });

  console.log(locale);
  console.log(translationFileContents);
  var bundle = new _fluent.FluentBundle(locale);
  bundle.addMessages((0, _fluent.ftl)([translationFileContents]));
  console.log(bundle.getMessage("h3527609692_ex-nome-sobrenome-nome-setor-nome"));
  var manipulator = {
    exit: function exit(path) {
      if (!(0, _ignoreAstPath.default)(path)) {
        var nodevalue = path.node.value.trim();
        if (nodevalue.indexOf("i18n:") === 0) nodevalue = nodevalue.substring("i18n:".length);
        var key = (0, _defineKey.default)(nodevalue);
        path.replaceWith(_core.types.stringLiteral(bundle.getMessage(key) || "???"));
        path.skip();
      }
    }
  };
  return {
    manipulateOptions: function manipulateOptions(opts, parserOpts) {
      ["jsx", "objectRestSpread", "dynamicImport", "classProperties"].forEach(function (pl) {
        return parserOpts.plugins.push(pl);
      });
    },
    visitor: {
      StringLiteral: manipulator,
      JSXText: manipulator
    }
  };
});

exports.default = _default;