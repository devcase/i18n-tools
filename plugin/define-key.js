"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _farmhash = _interopRequireDefault(require("farmhash"));

function pad(str, max) {
  str = str.toString();
  return str.length < max ? pad("0" + str, max) : str;
}

function _default(text) {
  var r = text.normalize('NFD').replace(/[\u0300-\u036f]/g, "").replace(/\W/g, "-").replace(/\-{2,}/g, "-").toLowerCase();
  if (r[0] === "-") r = r.substring(1);
  if (r[r.length - 1] === "-") r = r.substring(0, r.length - 1);
  return "h" + pad(_farmhash.default.hash32(text), 10) + "_" + r;
}