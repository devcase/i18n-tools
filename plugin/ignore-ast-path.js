"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function findParentWithType(path, type) {
  if (path.parent && path.parent.type === type) {
    return path.parent;
  }

  if (!path.parentPath) {
    return null;
  }

  return findParentWithType(path.parentPath, type);
}

function _default(path, options) {
  var ignoredRegex = [/[a-z][A-Z]/, /[a-zA-Z]\.[a-zA-Z]/, /^http/, /^\//, /\_/, /^\./, /^\#/, /px$/, /^[A-Z][A-Z][A-Z]$/, /^(accept|acceptCharset|accessKey|action|allowFullScreen|alt|async|autoComplete|autoFocus|autoPlay|capture|cellPadding|cellSpacing|challenge|charSet|checked|cite|classID|className|colSpan|cols|content|contentEditable|contextMenu|controls|controlsList|coords|crossOrigin|data|dateTime|default|defer|dir|disabled|download|draggable|encType|form|formAction|formEncType|formMethod|formNoValidate|formTarget|frameBorder|headers|height|hidden|high|href|hrefLang|htmlFor|httpEquiv|icon|id|inputMode|integrity|is|keyParams|keyType|kind|label|lang|list|loop|low|manifest|marginHeight|marginWidth|max|maxLength|media|mediaGroup|method|min|minLength|multiple|muted|name|noValidate|nonce|open|optimum|pattern|placeholder|poster|preload|profile|radioGroup|readOnly|rel|required|reversed|role|rowSpan|rows|sandbox|scope|scoped|scrolling|seamless|selected|shape|size|sizes|span|spellCheck|src|srcDoc|srcLang|srcSet|start|step|style|summary|tabIndex|target|title|type|useMap|value|width|wmode|wrap)$/];
  var requiredRegex = [/[a-zA-Z]/];
  var validJsxAttributes = ["label", "value", "aria-label", "title", "placeholder"];
  var validObjectPropertyKeys = ["text", "label", "labelPlural"];
  var value = path.node.value.trim();

  if (path.node.value.match(/^i18n:/)) {
    return false;
  }

  if (findParentWithType(path, "ImportDeclaration")) return true;
  var jsxAttributeParent = findParentWithType(path, "JSXAttribute");
  var jsxAttributeParentName = jsxAttributeParent && jsxAttributeParent.name.name;
  if (jsxAttributeParentName && !validJsxAttributes.find(function (t) {
    return t === jsxAttributeParentName;
  })) return true;

  if (findParentWithType(path, "CallExpression")) {
    return true;
  }

  var objectProperty;

  if (objectProperty = findParentWithType(path, "ObjectProperty")) {
    console.log(objectProperty.key);

    if (objectProperty.key === path.node) {
      return true;
    }

    if (!validObjectPropertyKeys.find(function (k) {
      return k === objectProperty.key.value || k === objectProperty.key.name;
    })) {
      return true;
    }
  }

  if (requiredRegex.find(function (regexp) {
    return !value.match(regexp);
  })) {
    return true;
  }

  if (ignoredRegex.find(function (regexp) {
    return value.match(regexp);
  })) {
    return true;
  }

  return false;
}