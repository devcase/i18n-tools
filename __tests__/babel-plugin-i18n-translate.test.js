import plugin from "../src/babel-plugin-i18n-translate";
import * as babel from "@babel/core";

it("works", () => {
  const example = `
    import React from 'react'

    export default function (props) {
        var value = {
            "label": "Incluir este texto",
            text: "Este texto também, mas sem as exclamações!!!!",
            "constantName": "Ignore this text",
            "error": "i18n:Erro de entrada"
        };
        return <div className="my-ignored-classname">
            <h1>     Meu texto internacionalizável!!!!</h1>
            <input type="text" placeholder="Nome do usuário"/>
        </div>
    }
`;

  const { code } = babel.transform(example, {
    plugins: [[plugin, { locale: "en-US" }]],
  });
  expect(code).toMatchSnapshot();
});

it.only("com apostofre", () => {
  const example = `
    const x = {
        label: 'E-mail do solicitante'
    }
`;

  const { code } = babel.transform(example, {
    configFile: false,
    plugins: [[plugin, { locale: "en-US" }]],
  });
  expect(code).toBe(`"use strict";

  var x = {
      label: 'Requestor\'s email'
  }
`);
});

it("espanol", () => {
  const example = `
    const x = {
        label: 'Últimas 24 horas'
    }
`;

  const { code } = babel.transform(example, {
    plugins: [[plugin, { locale: "es" }]],
  });
  expect(code).toBe(`"use strict";

  var x = {
      label: 'Últimas 24 horas'
  }
`);
});
