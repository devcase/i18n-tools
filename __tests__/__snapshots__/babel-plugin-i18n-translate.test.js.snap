// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"@babel/runtime/helpers/interopRequireDefault\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports[\\"default\\"] = _default;

var _react = _interopRequireDefault(require(\\"react\\"));

function _default(props) {
  var value = {
    \\"label\\": \\"Include this text\\",
    text: \\"This text also, but without exclamation points!!!!\\",
    \\"constantName\\": \\"Ignore this text\\",
    \\"error\\": \\"Input error\\"
  };
  return <div className=\\"my-ignored-classname\\">
            <h1>     My i18n text!!!!</h1>
            <input type=\\"text\\" placeholder=\\"Username\\" />
        </div>;
}"
`;
